Add extra content about bindings to advanced data manipulation chapter (#27)
arrowrbook/book: advanced_topics.qmd

| **Purpose**                    | Computes standard deviation                             | Computes standard deviation                                                             |
| **Handling of Missing Values** | `na.rm`: Boolean, removes `NA` values (default `FALSE`) | `skip_nulls`: Boolean, removes `NULL` values (default `FALSE`)                          |
| **Degrees of Freedom**         | Not available                                           | `ddof`: Integer, denominator degrees of freedom (default `0`)                           |
| **Minimum Count**              | Not available                                           | `min_count`: Integer, returns `NULL` if fewer non-null values are present (default `1`) |

There are also some differences in the options for these functions; while R's `sd` function has arguments `x`, the value to be passed in and `na.rm`, whether to remove `NA` values or not, Acero's stddev also has `ddof` - denominator degrees of freedom, `skip_nulls`, the equivalent of `na_rm`, and `min_count` - if there are fewer values than this, return `NULL`.

> This also highlights the slight different in terminology here; whereas R has separate concepts of `NA` and `NULL`, Arrow considers them to be the same things.
> 

Reviewer: ```suggestion
```

I don't think that's exactly right, and probably not worth trying to get precisely right either.

> You don't usually have to worry about these differences yourself though, as the code in the arrow R package does a direct translation from one to the other.

Reviewer: This would be a good place to do a query that has `sd(x)` in it, and then print the query, which prints the C++ option naming.

Author: Is this a bug or an unimplemented feature? Having difficulty constructing a query which can do this without me having to also introduce `show_query()`. Here's what I tried:

```r
open_dataset("data/person/") |> 
  summarise(sd_age = sd(AGEP))
```

gets me

```
FileSystemDataset (query)
sd_age: double

See $.data for the source Arrow object
```

I was wondering if it was just aggregations, but apparently not:

```r
open_dataset("data/person/") |> 
  transmute(sd_age = sd(AGEP))
```

gets 

```
FileSystemDataset (query)
sd_age: double

See $.data for the source Arrow object
```

I also tried doing some weird thing with filters which doesn't reeeeally make sense, but that doesn't work either:

```r
open_dataset("data/person/") |> 
  filter(AGEP > sd(AGEP))
```

```
Show in New Window
Error in AGEP > sd(AGEP) : 
  Expression not supported in filter() in Arrow
→ Call collect() first to pull data into R.
```

Author: Hmm, I think `sd` is special in some way here as if I swap it for `as.integer()` we get the output:

```
> open_dataset("data/person/") |> 
+     transmute(as.integer(AGEP))
FileSystemDataset (query)
as.integer(AGEP): int32 (cast(AGEP, {to_type=int32, allow_int_overflow=false, allow_time_truncate=false, allow_time_overflow=false, allow_decimal_truncate=false, allow_float_truncate=true, allow_invalid_utf8=false}))
```

Reviewer: `sd()` isn't special. Aggregations are just handled differently. The aggregation step happens and then there is a `collapse()`, so the thing you want is inside `$.data`. 

```
> query <- ds |> summarise(sd_age = sd(AGEP))
> query
FileSystemDataset (query)
sd_age: double

See $.data for the source Arrow object
> query$.data
FileSystemDataset (query)
AGEP: int32

* Aggregations:
sd_age: stddev(AGEP)
See $.data for the source Arrow object
```

But for whatever reason (well actually, see `format_aggregation()` in dplyr-summarize.R), aggregations don't print their options in the arrow_dplyr_query print method. But `show_query()` shows them.

```
> show_query(query)
ExecPlan with 4 nodes:
3:SinkNode{}
  2:ScalarAggregateNode{aggregates=[
	stddev(sd_age, {ddof=1, skip_nulls=false, min_count=0}),
]}
    1:ProjectNode{projection=["sd_age": AGEP]}
      0:SourceNode{}
>
```

This is "advanced topics" after all, so maybe `show_query()` isn't a bad thing to introduce. Or, use a scalar function instead of an aggregation as the example here. (Or both! We are short on pages!)

Author: Added!

Let's take a closer look at an example.

> Say we wanted to only return data for states whose names start with the letter "w".

Reviewer: Do you think we could come up with an example that isn't easily done with regular R functions? `utf8_swapcase` is a fun one that I don't think has an R/stringr version. `xor` is also interesting, if we had a non-contrived use case. There are other string functions that could be done with regular expressions but might be convenient not to, like `utf8_is_alnum`, `utf8_is_decimal`, `utf8_is_digit`, etc.

FWIW those probably also don't involve having to hunt down compute options, so that would make them easier to show. We could also show one that uses compute options, but unless we've had to add bindings for the function, we probably haven't wired up the options in `make_compute_options`, so it probably wouldn't just work anyway.

The options for the Acero function have been set based on the values which were passed in via the R expression.
We set the `startsWith()` argument `prefix` to "w", which was passed to the Acero `starts_with()` option called `pattern`.
As `startsWith()` is always case-sensitive, the Acero `starts_with()` option `ignore_case` has been set to `false`.

This approach is great for the most typical case where we know the R function we want to use as arrow automatically translates to the relevant Acero function, but what about if we just want to start with Acero functions?

We can view the full list of Acero functions available to use in arrow by calling `list_compute_functions()`.
At the time of writing there were 262 Acero functions available to use in R - the first 20 are shown below.

```{r}
#| label: list_compute_functions
list_compute_functions() |> 
  head(20)
```

You can also use these functions in your arrow dplyr pipelines by prefixing them with `"arrow_"`.

This isn't necessary when working with Acero functions which have a direct mapping to their R equivalents - which can be used instead. 

> However, there are some niche use-cases where this could be helpful - for example, if you want to use Acero options which don't exist in the R equivalent function.

Reviewer: This is true, but it would be more compelling of an example if it actually used an option that wasn't available through a regular binding.

Author: Good point!

We start off by considering which R function to use.
Let's go with the base R function `startsWith()`.
The table below shows a comparison with the Acero equivalent.

| Feature                     | R (`startsWith()`)                                | Acero (`starts_with()`)                            |
|-----------------|------------------------|--------------------------------|
| **Function Name**           | `startsWith`                                      | `starts_with`                                      |
| **Function Arguments**      | `prefix`                                     | `pattern`, `ignore_case = false`              |
| **Purpose**                 | Checks if a string starts with a specified prefix | Checks if a string starts with a specified pattern |
| **Prefix/Pattern Argument** | `prefix`: The prefix to match                     | `pattern`: The pattern to match                    |
| **Case Sensitivity**        | Always case-sensitive                             | `ignore_case`: Boolean, set to `false` by default  |

```{r}
#| label: startsWith
open_dataset("data/person/") |>
  select(location) |>
  filter(startsWith(x = location, prefix = "w"))
```

> The information about the Acero mapping is shown here in the line which starts with "\* Filter" - this contains a human-readable representation of the Arrow Expression created.

Reviewer: ```suggestion
The information about the Acero mapping is shown here in the line which starts with "\* Filter".
This contains a human-readable representation of the Arrow Expression created.
```

| **Degrees of Freedom**         | Not available                                           | `ddof`: Integer, denominator degrees of freedom (default `0`)                           |
| **Minimum Count**              | Not available                                           | `min_count`: Integer, returns `NULL` if fewer non-null values are present (default `1`) |

There are also some differences in the options for these functions; while R's `sd` function has arguments `x`, the value to be passed in and `na.rm`, whether to remove `NA` values or not, Acero's stddev also has `ddof` - denominator degrees of freedom, `skip_nulls`, the equivalent of `na_rm`, and `min_count` - if there are fewer values than this, return `NULL`.

```{r}
#| label: sd_binding
open_dataset("data/person/") |>
  summarise(sd_age = sd(AGEP))
```

The output here doesn't show the function mapping as aggregation functions are handled differently from `select()`, `filter()`, and `mutate()`.
The aggregation step is contained in the nested `.data` attribute of the query.
This nesting ensures that the steps of the query are evaluated in this order.

> While `select()`, `filter()`, and `mutate()` can happen successively and be reflected in a single query step, any actions that happen on the data after `summarize()` (and, for that matter, joins) cannot be collapsed or rearranged like that.

Reviewer: I added this paragraph, not sure it's totally clear. LMK if you want me to expand on it: could discuss in the "query optimizer" section above how the flattening of select/filter/mutate works, so then this would call back to that.

Also, should we introduce `collapse()`? If there were cases where that flattening were suboptimal, like if it led Acero to recompute a complex expression many times instead of doing it once and then mutating the result, it could be helpful. Though I haven't done the analysis to find a case where it would help.

Reviewer: I made https://github.com/apache/arrow/issues/43894 to make this print better (if you looked at `.data`), seems easy enough, if we care.

Author: I guess we could be more explicit about exact why we can't just collapse or rearrange them - I think it's obvious to us but not necessarily to others.

We'll also explore the converse of this: functions that exist in Acero and don't have R equivalents, and how you can access these in your arrow pipelines.
Then, we're going to show how you can call back from Acero into R to do calculations where functions you want to use inside don't have a corresponding Acero function.
For other cases, we'll show how the Arrow format's focus on interoperability lets you stream data to and from other query engines, such as DuckDB, to access different functionality they have.
Finally, we'll look at a practical case of extending Arrow's type system to work with geospatial data.

## Acero

Acero is the Arrow C++ library's query execution engine.
A query execution engine is a library which allows users to perform queries against a data source.
It takes input data and the operations to be performed on the data, combines them, and returns the output from this process.

It typically has multiple steps, which could include things like:

* taking the query and parsing it into an algebraic format
* re-ordering and optimising the query in order to run it in the most efficient way[^optimizer]
* actually running the query

> [^optimizer]: Acero doesn't have a logical query optimizer, but as we have seen, it does support optimizations like predicated pushdown: this is what makes filtering on partitioned datasets fast. To take advantage of these optimzations, there is code in the R package that reorders queries before passing it to Acero.

Author: We should either spell out what "predicate pushdown" is or wrap it in quotes, as it's a super specialist term

> [^optimizer]: Acero doesn't have a logical query optimizer, but as we have seen, it does support optimizations like predicated pushdown: this is what makes filtering on partitioned datasets fast. To take advantage of these optimzations, there is code in the R package that reorders queries before passing it to Acero.

Author: I don't think there's any harm in it!

```{r}
#| label: minus_binding
open_dataset("data/person/") |>
  transmute(age_previous_year = AGEP - 1)
```

When we print the query, the output shows that new column `age_previous_year` is calculated by calling the Acero function `subtract_checked()` and passing in a reference to the `AGEP` column and the number `1`.

### Bindings for more complex functions

Mappings also exist for slightly more complex functions that have additional parameters.

Say we wanted to only return data for states whose names start with the letter "w".
We can take our dataset and use `dplyr::filter()` to return this subset, but we need to pick a function to pass to `dplyr::filter()` to actually *do* the filtering.

Let's start first by looking at the Acero function we want Arrow to use.
To be clear: we wouldn't usually have to do this when working with arrow.
This is just to illustrate how this all works.

If we look at the [Acero compute function docs](https://arrow.apache.org/docs/cpp/compute.html#available-functions), we'll see that there is a function `starts_with` which does this.

> It has 2 options: `pattern` the substring to match at the start of the string, and `ignore_case`, whether the matching should be case sensitive.

Author: ```suggestion
It has 2 options: `pattern`, the substring to match at the start of the string, and `ignore_case`, whether the matching should be case sensitive.
```

> It has 2 options - `pattern` the substring to match at the start of the string, and `ignore_case`, whether the matching should be case sensitive.

Author: ```suggestion
It has 2 options: `pattern` the substring to match at the start of the string, and `ignore_case`, whether the matching should be case sensitive.
```

We'll also explore the converse of this: functions that exist in Acero and don't have R equivalents, and how you can access these in your arrow pipelines.
Then, we're going to show how you can call back from Acero into R to do calculations where functions you want to use inside don't have a corresponding Acero function.
For other cases, we'll show how the Arrow format's focus on interoperability lets you stream data to and from other query engines, such as DuckDB, to access different functionality they have.
Finally, we'll look at a practical case of extending Arrow's type system to work with geospatial data.

## Acero

Acero is the Arrow C++ library's query execution engine.
A query execution engine is a library which allows users to perform queries against a data source.
It takes input data and the operations to be performed on the data, combines them, and returns the output from this process.

It typically has multiple steps, which could include things like:

* taking the query and parsing it into an algebraic format
* re-ordering and optimising the query in order to run it in the most efficient way[^optimizer]
* actually running the query

> [^optimizer]: Acero doesn't have a logical query optimizer, but as we have seen, it does support optimizations like predicated pushdown: this is what makes filtering on partitioned datasets fast. To take advantage of these optimzations, there is code in the R package that reorders queries before passing it to Acero for most common workflows.

Author: ```suggestion
[^optimizer]: Acero doesn't have a logical query optimizer, but as we have seen, it does support optimizations like "predicate pushdown" - filtering the data where it is stored rather than when in memory, which is what makes filtering on partitioned datasets fast. To take advantage of these optimizations, there is code in the R package that reorders queries before passing it to Acero for most common workflows.
```
